File: host/src/main/java/com/example/testapp/widget/XfermodeRoundImageView.java
package com.example.testapp.widget;

import android.content.Context;
import android.graphics.*;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;

import androidx.appcompat.widget.AppCompatImageView;

import android.util.AttributeSet;
import android.util.Log;

/**
 * Created by anxiao on 2017/8/11.
 * xfermode round image
 */

public class XfermodeRoundImageView extends AppCompatImageView {

    private String tag = "XfermodeRoundImageView";

    private Paint mPaint = new Paint();
    private PorterDuffXfermode porterDuffXfermode = new PorterDuffXfermode(PorterDuff.Mode.SRC_IN);

    public XfermodeRoundImageView(Context context) {
        super(context);

    }

    public XfermodeRoundImageView(Context context, AttributeSet attrs) {
        super(context, attrs);

    }



    @Override
    public void onDraw(Canvas canvas) {



        int width = getWidth() - getPaddingLeft() - getPaddingRight();
        int height = getHeight() - getPaddingTop() - getPaddingBottom();

        Bitmap cBitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);
        Canvas xCanvas = new Canvas(cBitmap);

        Drawable drawable = getDrawable();

        if (!(drawable instanceof BitmapDrawable)) {
            try {
                throw new Exception("drawable isn't BitmapDrawable");
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

        Bitmap bitmap = scale(((BitmapDrawable) drawable).getBitmap());

        float left = 0F;
        float top = 0F;

        if (bitmap.getWidth() > bitmap.getHeight()) {
            left = ((width - bitmap.getWidth()) / 2F);
        } else {
            top = ((height - bitmap.getHeight()) / 2F);
        }

        float radius = width > height ? height / 2F : width / 2F;


        mPaint.setAntiAlias(true);

        xCanvas.drawCircle(radius, radius, radius, mPaint);

        xCanvas.drawBitmap(cBitmap, 0F, 0F, mPaint);

        mPaint.setXfermode(porterDuffXfermode);

        xCanvas.drawBitmap(bitmap, left, top, mPaint);

        mPaint.setXfermode(null);

        canvas.drawBitmap(cBitmap, 0F, 0F, mPaint);

    }


    private Bitmap scale(Bitmap bitmap) {

        float cWidth = getWidth() - getPaddingLeft() - getPaddingRight();
        float cHeight = getHeight() - getPaddingTop() - getPaddingBottom();
        int oWidth = bitmap.getWidth();
        int oHeight = bitmap.getHeight();

        Matrix matrix = new Matrix();

        float temp = oWidth > oHeight ? cHeight / oHeight : cWidth / oWidth;

        matrix.postScale(temp, temp);

        return Bitmap.createBitmap(bitmap, 0, 0, oWidth, oHeight, matrix, false);
    }

}

File: app/src/main/java/com/example/testapp/widget/XfermodeRoundImageView.kt
package com.example.testapp.mytestapplication.customer_drawable

import android.content.Context
import android.graphics.*

import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable

import androidx.appcompat.widget.AppCompatImageView
import android.util.AttributeSet

/**
 * Created by anxiao on 2017/8/11.
 * xfermode round image
 */

class XfermodeRoundImageView : AppCompatImageView {

    private val mPaint = Paint()

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)

    private val porterDuffXfermode = PorterDuffXfermode(PorterDuff.Mode.SRC_IN)


    val cBitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)



    override fun onDraw(canvas: Canvas) {


        val width = width - paddingLeft - paddingRight
        val height = height - paddingTop - paddingBottom

        if (drawable !is BitmapDrawable) {
            throw Exception("drawable is not bitmap")
        }

        val bitmap = scale((drawable as BitmapDrawable).bitmap)

        var left = 0F
        var top = 0F

        if (bitmap.width > bitmap.height) {

            left = ((width - bitmap.width) / 2).toFloat()

        } else {

            top = ((height - bitmap.height) / 2).toFloat()

        }

        val radius = if (width > height)
            height / 2
        else
            width / 2


        mPaint.isAntiAlias = true

        val cBitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        val xCanvas = Canvas(cBitmap)

        xCanvas.drawCircle(radius.toFloat(), radius.toFloat(), radius.toFloat(), mPaint)

        xCanvas.drawBitmap(cBitmap, 0F, 0F, mPaint)

        mPaint.xfermode = porterDuffXfermode

        xCanvas.drawBitmap(bitmap, left, top, mPaint)

        mPaint.xfermode = null

        canvas.drawBitmap(cBitmap, 0F, 0F, mPaint)

    }


    private fun scale(bitmap: Bitmap): Bitmap {

        val cWidth = width - paddingLeft - paddingRight
        val cHeight = height - paddingTop - paddingBottom

        val oWidth = bitmap.width
        val oHeight = bitmap.height

        val matrix = Matrix()
        val temp = if (oWidth > oHeight) {
            cHeight.toFloat() / oHeight
        } else {
            cWidth.toFloat() / oWidth
        }
        matrix.postScale(temp, temp)
        return Bitmap.createBitmap(bitmap, 0, 0, oWidth, oHeight, matrix, false)
    }

}
